--- a/drivers/mmc/core/core.c
+++ b/drivers/mmc/core/core.c
@@ -1140,7 +1140,7 @@ u32 mmc_select_voltage(struct mmc_host *
 
 	ocr &= host->ocr_avail;
 	if (!ocr) {
-		dev_warn(mmc_dev(host), "no support for card's volts\n");
+		dev_warn(mmc_dev(host), "no card detected\n");
 		return 0;
 	}
 
--- a/drivers/mmc/host/mmc_spi.c
+++ b/drivers/mmc/host/mmc_spi.c
@@ -1415,6 +1415,9 @@ static int mmc_spi_probe(struct spi_devi
 		mmc->caps2 |= host->pdata->caps2;
 	}
 
+	if ((status = mmc_of_parse(mmc)))
+		goto fail_gpiod_request;
+
 	status = mmc_add_host(mmc);
 	if (status != 0)
 		goto fail_glue_init;
